File: crates/alkanes-runtime/src/runtime.rs

#[allow(unused_imports)]
use crate::imports::{
    __balance, __call, __delegatecall, __fuel, __height, __load_block, __load_context,
    __load_storage, __load_transaction, __log, __request_block, __request_context,
    __request_storage, __request_transaction, __returndatacopy, __sequence, __staticcall,
    abort, /*, __load_output, __request_output */
};
use crate::storage::StoragePointer;
#[allow(unused_imports)]
use crate::{
    println,
    stdio::{stdout, Write},
};
use anyhow::{anyhow, Result};
use bitcoin::{block::Header, Transaction, Txid};
#[allow(unused_imports)]
use metashrew_support::compat::{to_arraybuffer_layout, to_passback_ptr, to_ptr};
use metashrew_support::{index_pointer::KeyValuePointer, utils::consensus_decode};
use std::io::Cursor;

#[cfg(feature = "panic-hook")]
use crate::compat::panic_hook;

#[allow(unused_imports)]
use alkanes_support::{
    cellpack::Cellpack,
    context::Context,
    id::AlkaneId,
    parcel::{AlkaneTransfer, AlkaneTransferParcel},
    response::{CallResponse, ExtendedCallResponse},
    storage::StorageMap,
};
#[cfg(feature = "panic-hook")]
use std::panic;

fn _abort() {
    unsafe {
        abort(0, 0, 0, 0);
    }
}



#[allow(static_mut_refs)]
pub fn handle_success(response: CallResponse, env: &mut AlkaneEnvironment) -> ExtendedCallResponse {
    let mut extended: ExtendedCallResponse = response.into();
    let storage = StorageMap(
        env.cache
            .iter()
            .map(|(k, v)| ((**k).clone(), (**v).clone()))
            .collect(),
    );
    extended.storage = storage;
    extended
}

pub fn handle_error(error: &str) -> ExtendedCallResponse {
    let mut response = CallResponse::default();
    let mut data: Vec<u8> = vec![0x08, 0xc3, 0x79, 0xa0];
    data.extend(error.as_bytes());
    response.data = data;
    _abort();
    response.into()
}

pub fn prepare_response(response: ExtendedCallResponse) -> Vec<u8> {
    response.serialize()
}

pub fn response_to_i32(response: ExtendedCallResponse) -> i32 {
    let serialized = prepare_response(response);
    let response_bytes = to_arraybuffer_layout(&serialized);
    Box::leak(Box::new(response_bytes)).as_mut_ptr() as usize as i32 + 4
}

pub trait Extcall {
    fn __call(cellpack: i32, outgoing_alkanes: i32, checkpoint: i32, fuel: u64) -> i32;
    #[allow(static_mut_refs)]
    fn call(
        env: &mut AlkaneEnvironment,
        cellpack: &Cellpack,
        outgoing_alkanes: &AlkaneTransferParcel,
        fuel: u64,
    ) -> Result<CallResponse> {
        let mut cellpack_buffer = to_arraybuffer_layout::<&[u8]>(&cellpack.serialize());
        let mut outgoing_alkanes_buffer: Vec<u8> =
            to_arraybuffer_layout::<&[u8]>(&outgoing_alkanes.serialize());
        let storage = StorageMap(
            env.cache
                .iter()
                .map(|(k, v)| ((**k).clone(), (**v).clone()))
                .collect(),
        );
        let mut storage_map_buffer = to_arraybuffer_layout::<&[u8]>(&storage.serialize());
        let _call_result = Self::__call(
            to_passback_ptr(&mut cellpack_buffer),
            to_passback_ptr(&mut outgoing_alkanes_buffer),
            to_passback_ptr(&mut storage_map_buffer),
            fuel,
        );
        if _call_result < 0 {
            let call_result = _call_result.abs() as usize;
            let mut returndata = to_arraybuffer_layout(&vec![0; call_result]);
            unsafe {
                __returndatacopy(to_passback_ptr(&mut returndata));
            }
            if returndata.len() < 20 {
                return Err(anyhow!(format!(
                    "Extcall failed, and returndatacopy len ({}) < AlkanesTransferParcel min size 20 ",
                    returndata.len()
                )));
            }
            let response = CallResponse::parse(&mut Cursor::new((&returndata[4..]).to_vec()))?;
            if response.data.len() <= 4 || &response.data[0..4] != &[0x08, 0xc3, 0x79, 0xa0] {
                return Err(anyhow!("Extcall failed (no details available)"));
            }
            let error_message = String::from_utf8_lossy(&response.data[4..]).to_string();
            return Err(anyhow!("Extcall failed: {}", error_message));
        } else {
            let call_result = _call_result as usize;
            let mut returndata = to_arraybuffer_layout(&vec![0; call_result]);
            unsafe {
                __returndatacopy(to_passback_ptr(&mut returndata));
            }
            if returndata.len() < 20 {
                return Err(anyhow!(format!(
                    "Extcall succeeded, but returndatacopy len ({}) < AlkanesTransferParcel min size 20 ",
                    returndata.len()
                )));
            }
            let response = CallResponse::parse(&mut Cursor::new((&returndata[4..]).to_vec()))?;
            Ok(response)
        }
    }
}

pub struct Call(());

impl Extcall for Call {
    fn __call(cellpack: i32, outgoing_alkanes: i32, checkpoint: i32, fuel: u64) -> i32 {
        unsafe { __call(cellpack, outgoing_alkanes, checkpoint, fuel) }
    }
}

pub struct Delegatecall(());

impl Extcall for Delegatecall {
    fn __call(cellpack: i32, outgoing_alkanes: i32, checkpoint: i32, fuel: u64) -> i32 {
        unsafe { __delegatecall(cellpack, outgoing_alkanes, checkpoint, fuel) }
    }
}

pub struct Staticcall(());

impl Extcall for Staticcall {
    fn __call(cellpack: i32, outgoing_alkanes: i32, checkpoint: i32, fuel: u64) -> i32 {
        unsafe { __staticcall(cellpack, outgoing_alkanes, checkpoint, fuel) }
    }
}

pub use crate::environment::AlkaneEnvironment;

pub trait AlkaneResponder: 'static {
    fn env(&mut self) -> &mut AlkaneEnvironment;
    fn observe_initialization(&mut self) -> Result<()> {
        let mut pointer: StoragePointer = KeyValuePointer::<AlkaneEnvironment>::from_keyword("/initialized");
        if pointer.get(self.env()).len() == 0 {
            pointer.set_value::<u8>(self.env(), 0x01);
            Ok(())
        } else {
            Err(anyhow!("already initialized"))
        }
    }
    fn observe_proxy_initialization(&mut self) -> Result<()> {
        let mut pointer: StoragePointer = KeyValuePointer::<AlkaneEnvironment>::from_keyword("/proxy_initialized");
        if pointer.get(self.env()).len() == 0 {
            pointer.set_value::<u8>(self.env(), 0x01);
            Ok(())
        } else {
            Err(anyhow!("proxy already initialized"))
        }
    }
    fn context(&self) -> Result<Context> {
        unsafe {
            let mut buffer: Vec<u8> = to_arraybuffer_layout(vec![0; __request_context() as usize]);
            __load_context(to_ptr(&mut buffer) + 4);
            let res = Context::parse(&mut Cursor::<Vec<u8>>::new((&buffer[4..]).to_vec()));
            res
        }
    }
    fn block(&self) -> Vec<u8> {
        unsafe {
            let mut buffer: Vec<u8> = to_arraybuffer_layout(vec![0; __request_block() as usize]);
            __load_block(to_ptr(&mut buffer) + 4);
            (&buffer[4..]).to_vec()
        }
    }

    fn transaction(&self) -> Vec<u8> {
        unsafe {
            let mut buffer: Vec<u8> =
                to_arraybuffer_layout(vec![0; __request_transaction() as usize]);
            __load_transaction(to_ptr(&mut buffer) + 4);
            (&buffer[4..]).to_vec()
        }
    }
    fn transaction_object(&self) -> Result<Transaction> {
        Ok(consensus_decode::<Transaction>(&mut std::io::Cursor::new(
            self.transaction(),
        ))?)
    }
    fn transaction_id(&self) -> Result<Txid> {
        Ok(self.transaction_object()?.compute_txid())
    }
    /*
    fn output(&self, v: &OutPoint) -> Result<Vec<u8>> {
        let mut buffer = to_arraybuffer_layout(consensus_encode(v)?);
        let serialized = to_passback_ptr(&mut buffer);
        unsafe {
            let mut result: Vec<u8> =
                to_arraybuffer_layout(vec![0; __request_output(serialized) as usize]);
            let sz = __load_output(serialized, to_passback_ptr(&mut result));
            if sz == i32::MAX {
              Err(anyhow!("error fetching output"))
            } else if sz == 0 {
              Err(anyhow!("output not found"))
            } else {
              Ok((&result[4..]).to_vec())
            }
        }
    }
    */

    fn balance(&self, who: &AlkaneId, what: &AlkaneId) -> u128 {
        unsafe {
            let mut who_bytes: Vec<u8> = to_arraybuffer_layout::<Vec<u8>>(who.clone().into());
            let mut what_bytes: Vec<u8> = to_arraybuffer_layout::<Vec<u8>>(what.clone().into());
            let who_ptr = to_ptr(&mut who_bytes) + 4;
            let what_ptr = to_ptr(&mut what_bytes) + 4;
            let mut output: Vec<u8> = to_arraybuffer_layout::<Vec<u8>>(vec![0u8; 16]);
            __balance(who_ptr, what_ptr, to_ptr(&mut output) + 4);
            u128::from_le_bytes((&output[4..]).try_into().unwrap())
        }
    }
    fn sequence(&self) -> u128 {
        unsafe {
            let mut buffer: Vec<u8> = to_arraybuffer_layout(vec![0; 16]);
            __sequence(to_ptr(&mut buffer) + 4);
            u128::from_le_bytes((&buffer[4..]).try_into().unwrap())
        }
    }
    fn fuel(&self) -> u64 {
        unsafe {
            let mut buffer: Vec<u8> = to_arraybuffer_layout(vec![0; 8]);
            __fuel(to_ptr(&mut buffer) + 4);
            u64::from_le_bytes((&buffer[4..]).try_into().unwrap())
        }
    }
    fn height(&self) -> u64 {
        unsafe {
            let mut buffer: Vec<u8> = to_arraybuffer_layout(vec![0; 8]);
            __height(to_ptr(&mut buffer) + 4);
            u64::from_le_bytes((&buffer[4..]).try_into().unwrap())
        }
    }
    fn extcall<T: Extcall>(
        &mut self,
        cellpack: &Cellpack,
        outgoing_alkanes: &AlkaneTransferParcel,
        fuel: u64,
    ) -> Result<CallResponse> {
        T::call(self.env(), cellpack, outgoing_alkanes, fuel)
    }
    fn call(
        &mut self,
        cellpack: &Cellpack,
        outgoing_alkanes: &AlkaneTransferParcel,
        fuel: u64,
    ) -> Result<CallResponse> {
        self.extcall::<Call>(cellpack, outgoing_alkanes, fuel)
    }
    fn delegatecall(
        &mut self,
        cellpack: &Cellpack,
        outgoing_alkanes: &AlkaneTransferParcel,
        fuel: u64,
    ) -> Result<CallResponse> {
        self.extcall::<Delegatecall>(cellpack, outgoing_alkanes, fuel)
    }
    fn staticcall(
        &mut self,
        cellpack: &Cellpack,
        outgoing_alkanes: &AlkaneTransferParcel,
        fuel: u64,
    ) -> Result<CallResponse> {
        self.extcall::<Staticcall>(cellpack, outgoing_alkanes, fuel)
    }

    fn block_header(&mut self) -> Result<Header> {
        let result = self.staticcall(
            &Cellpack {
                target: AlkaneId {
                    block: 800000000,
                    tx: 0,
                },
                inputs: vec![],
            },
            &AlkaneTransferParcel::default(),
            self.fuel(),
        )?;
        consensus_decode::<Header>(&mut std::io::Cursor::new(result.data))
    }

    fn coinbase_tx(&mut self) -> Result<Transaction> {
        let result = self.staticcall(
            &Cellpack {
                target: AlkaneId {
                    block: 800000000,
                    tx: 1,
                },
                inputs: vec![],
            },
            &AlkaneTransferParcel::default(),
            self.fuel(),
        )?;
        consensus_decode::<Transaction>(&mut std::io::Cursor::new(result.data))
    }

    fn number_diesel_mints(&mut self) -> Result<u128> {
        let result = self.staticcall(
            &Cellpack {
                target: AlkaneId {
                    block: 800000000,
                    tx: 2,
                },
                inputs: vec![],
            },
            &AlkaneTransferParcel::default(),
            self.fuel(),
        )?;
        Ok(u128::from_le_bytes(result.data[0..16].try_into()?))
    }

    fn total_miner_fee(&mut self) -> Result<u128> {
        let result = self.staticcall(
            &Cellpack {
                target: AlkaneId {
                    block: 800000000,
                    tx: 3,
                },
                inputs: vec![],
            },
            &AlkaneTransferParcel::default(),
            self.fuel(),
        )?;
        Ok(u128::from_le_bytes(result.data[0..16].try_into()?))
    }

    /// Fallback function that gets called when an opcode is not recognized
    ///
    /// This default implementation reverts with an error.
    /// Contracts can override this method to provide custom fallback behavior.
    fn fallback(&mut self) -> Result<CallResponse> {
        Err(anyhow!("Unrecognized opcode"))
    }
}

File: crates/alkanes-cli-common/src/alkanes/asc/alkanes-asm-common/assembly/alkanes/runtime.ts
// Alkanes runtime host function imports
// All pointer parameters point to ArrayBuffer data (length is at ptr-4)

/**
 * Abort execution with error
 * @param a Message pointer (or 0)
 * @param b Message length (or 0)  
 * @param c Line number (or 0)
 * @param d Column number (or 0)
 */
// @ts-ignore: decorator
@external("env", "abort")
export declare function abort(a: i32, b: i32, c: i32, d: i32): void;

/**
 * Load a value from storage
 * @param k Key pointer (ArrayBuffer)
 * @param v Output pointer (ArrayBuffer) to write value
 * @returns Length of value written
 */
// @ts-ignore: decorator
@external("env", "__load_storage")
export declare function __load_storage(k: i32, v: i32): i32;

/**
 * Request the size of a storage value
 * @param k Key pointer (ArrayBuffer)
 * @returns Size of the value in bytes
 */
// @ts-ignore: decorator
@external("env", "__request_storage")
export declare function __request_storage(k: i32): i32;

/**
 * Log a message (for debugging)
 * @param v Message pointer (ArrayBuffer)
 */
// @ts-ignore: decorator
@external("env", "__log")
export declare function __log(v: i32): void;

/**
 * Get balance of an alkane
 * @param who Address pointer (AlkaneId ArrayBuffer)
 * @param what Alkane ID pointer (AlkaneId ArrayBuffer)
 * @param output Output pointer (ArrayBuffer) to write balance (16 bytes)
 */
// @ts-ignore: decorator
@external("env", "__balance")
export declare function __balance(who: i32, what: i32, output: i32): void;

/**
 * Request the size of the execution context
 * @returns Size of the context in bytes
 */
// @ts-ignore: decorator
@external("env", "__request_context")
export declare function __request_context(): i32;

/**
 * Load the execution context into memory at ptr
 * @param output Pointer to write context data (ArrayBuffer)
 * @returns The pointer (for chaining)
 */
// @ts-ignore: decorator
@external("env", "__load_context")
export declare function __load_context(output: i32): i32;

/**
 * Get the current sequence number
 * @param output Output pointer (ArrayBuffer) to write sequence (16 bytes u128)
 */
// @ts-ignore: decorator
@external("env", "__sequence")
export declare function __sequence(output: i32): void;

/**
 * Get the remaining fuel
 * @param output Output pointer (ArrayBuffer) to write fuel (8 bytes u64)
 */
// @ts-ignore: decorator
@external("env", "__fuel")
export declare function __fuel(output: i32): void;

/**
 * Get the current block height
 * @param output Output pointer (ArrayBuffer) to write height (8 bytes u64)
 */
// @ts-ignore: decorator
@external("env", "__height")
export declare function __height(output: i32): void;

/**
 * Copy return data from last call into memory at ptr
 * @param output Pointer to write return data (ArrayBuffer)
 */
// @ts-ignore: decorator
@external("env", "__returndatacopy")
export declare function __returndatacopy(output: i32): void;

/**
 * Request the size of the current transaction
 * @returns Size of the transaction in bytes
 */
// @ts-ignore: decorator
@external("env", "__request_transaction")
export declare function __request_transaction(): i32;

/**
 * Load the current transaction into memory
 * @param output Pointer to write transaction data (ArrayBuffer)
 */
// @ts-ignore: decorator
@external("env", "__load_transaction")
export declare function __load_transaction(output: i32): void;

/**
 * Request the size of the current block
 * @returns Size of the block in bytes
 */
// @ts-ignore: decorator
@external("env", "__request_block")
export declare function __request_block(): i32;

/**
 * Load the current block into memory
 * @param output Pointer to write block data (ArrayBuffer)
 */
// @ts-ignore: decorator
@external("env", "__load_block")
export declare function __load_block(output: i32): void;

/**
 * Make a regular call to another alkane
 * @param cellpack Pointer to cellpack data (ArrayBuffer)
 * @param incoming_alkanes Pointer to AlkaneTransferParcel (ArrayBuffer)
 * @param checkpoint Pointer to StorageMap (ArrayBuffer)
 * @param start_fuel Maximum fuel to use
 * @returns Length of return data if >= 0, error code if < 0
 */
// @ts-ignore: decorator
@external("env", "__call")
export declare function __call(
  cellpack: i32,
  incoming_alkanes: i32,
  checkpoint: i32,
  start_fuel: u64
): i32;

/**
 * Make a staticcall to another alkane (read-only, cannot modify state)
 * @param cellpack Pointer to cellpack data (ArrayBuffer)
 * @param incoming_alkanes Pointer to AlkaneTransferParcel (ArrayBuffer)
 * @param checkpoint Pointer to StorageMap (ArrayBuffer)
 * @param start_fuel Maximum fuel to use
 * @returns Length of return data if >= 0, error code if < 0
 */
// @ts-ignore: decorator
@external("env", "__staticcall")
export declare function __staticcall(
  cellpack: i32,
  incoming_alkanes: i32,
  checkpoint: i32,
  start_fuel: u64
): i32;

/**
 * Make a delegatecall to another alkane (runs in current context)
 * @param cellpack Pointer to cellpack data (ArrayBuffer)
 * @param incoming_alkanes Pointer to AlkaneTransferParcel (ArrayBuffer)
 * @param checkpoint Pointer to StorageMap (ArrayBuffer)
 * @param start_fuel Maximum fuel to use
 * @returns Length of return data if >= 0, error code if < 0
 */
// @ts-ignore: decorator
@external("env", "__delegatecall")
export declare function __delegatecall(
  cellpack: i32,
  incoming_alkanes: i32,
  checkpoint: i32,
  start_fuel: u64
): i32;
